### DISCUSSION TITLE
Creating or accessing column value indexes

### DISCUSSION THREAD

**Discussion by skokenes** [duckdb/duckdb#6711] [Q&A] [answered]
For some data analytics and data science algorithms, it is useful to have indexed representations of a data table. For example, assume a simple table in a CSV like

| Region           | Product          | Sales |
| ------------- | ------------- | --- |
| North  | Shoes  | 100 |
| North  | Shirts | 50 |
| South  | Shoes  | 25 |
| South  | Pants | 50 |

For certain algorithms, I want to look at the table with integer indexes replacing the column values like so:
| Region           | Product          | Sales |
| ------------- | ------------- | --- |
| 0  | 0  | 0 |
| 0  | 1 | 1 |
| 1  | 0  | 2 |
| 1  | 2 | 1 |

I have a couple of questions about working with data like this in DuckDB:

1. The docs state that "A [min-max index](https://en.wikipedia.org/wiki/Block_Range_Index) is automatically created for columns of all [general-purpose data types](https://duckdb.org/docs/sql/data_types/overview).". If that is the case, is it possible to expose functions that allow us access index values from SQL statements? Then I could load my data as is but still use the numeric index representations when needed in formulas. I am imagining functions like
```sql
get_index('Region', 'North') --- returns 0

get_value('Region', 0) --- returns 'North'
```
Maybe even ways to query the index as if it was a table? Like `SELECT * FROM index(table, colName)`
2. If the above isn't possible or not worth consideration for the roadmap, are there any recommendations on how to create such a data representation in DuckDB just using SQL?

**Comment by Mytherin** [comment, marked as answer]
I think what you are looking for is a dictionary representation of the data, which can be done using [enum types](https://duckdb.org/docs/sql/data_types/enum).

For example, in your example:

```sql
D CREATE TABLE tbl(Region VARCHAR, Product VARCHAR, Sales INT);
D INSERT INTO tbl VALUES ('North', 'Shoes', 100);
D INSERT INTO tbl VALUES ('North', 'Shirts', 50);
D INSERT INTO tbl VALUES ('South', 'Shoes', 25);
D INSERT INTO tbl VALUES ('South', 'Pants', 50);
-- create an enum type for the Region column
D CREATE TYPE region_enum AS ENUM(SELECT Region FROM tbl);
-- select all possible values stored in the enum
D SELECT enum_range(NULL::region_enum) AS all_values;
┌────────────────┐
│   all_values   │
│   varchar[]    │
├────────────────┤
│ [North, South] │
└────────────────┘
-- select the corresponding enum_code for the enum
D SELECT Region, enum_code(Region::region_enum) AS region_code FROM tbl;
┌─────────┬─────────────┐
│ Region  │ region_code │
│ varchar │    uint8    │
├─────────┼─────────────┤
│ North   │           0 │
│ North   │           0 │
│ South   │           1 │
│ South   │           1 │
└─────────┴─────────────┘
```

**Reply by skokenes** [reply]
Thanks for the quick reply. I've considered enums before but I am not sure they are suitable for my scenario for a few reasons
1. They only support VARCHAR, where as I want an index for every column and generic data type. So for example, the Sales column needs encoding as well
2. I have seen in some Issues in the duckdb repo that you've said that enums should be avoided for large distinct sets of values; I believe I saw reference to <10k values, but not sure. The tables that I am loading could have a very high amount of unique values in a column

Other options?

**Reply by Mytherin** [reply]
1) With a cast you can use enums with any type:

```sql
CREATE TYPE sales_enum AS ENUM(SELECT Sales::VARCHAR FROM tbl);
SELECT Sales, enum_code(Sales::VARCHAR::sales_enum) AS region_code FROM tbl;
┌───────┬─────────────┐
│ Sales │ region_code │
│ int32 │    uint8    │
├───────┼─────────────┤
│   100 │           0 │
│    50 │           1 │
│    25 │           2 │
│    50 │           1 │
└───────┴─────────────┘
```

2) Enums are indeed not a good data type when you have millions of unique values on the storage front, but this is not using enums to store the data.

Another option is to make a look-up table using `GROUP BY` and `row_number`, and do a join to fetch the codes, e.g.:

```sql
CREATE TABLE region_codes AS SELECT Region, row_number() OVER () - 1 as val FROM (SELECT Region FROM tbl GROUP BY Region);
FROM region_codes;
┌─────────┬───────┐
│ Region  │  val  │
│ varchar │ int64 │
├─────────┼───────┤
│ North   │     0 │
│ South   │     1 │
└─────────┴───────┘
SELECT Region, region_codes.val AS region_val FROM tbl JOIN region_codes USING (Region);
┌─────────┬────────────┐
│ Region  │ region_val │
│ varchar │   int64    │
├─────────┼────────────┤
│ North   │          0 │
│ North   │          0 │
│ South   │          1 │
│ South   │          1 │
└─────────┴────────────┘

```

**Reply by skokenes** [reply]
Thanks, I will experiment with using the enums with casting

**Reply by skokenes** [reply]
Just a follow up question...what if I DID want to store these dictionaries in my db so that I didn't have to recreate them every time I am analyzing the same set of data? With the enum approach, do I then need to be concerned about storage impact?

**Reply by Mytherin** [reply]
If you only store the type itself then you should be fine. We have also recently fixed some issues around storage of large enums so even if you store the data columns **as** the enum types the impact should not be that big anymore.
